theseus/labs/lie/lie_tensor.py

```diff
@@ -0,0 +1,112 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from enum import Enum
+from typing import List, Tuple, Union
+
+import torch
+
+from theseus.labs.lie.functional.lie_group import UnaryOperatorOpFnType
+from theseus.labs.lie.functional import se3 as _se3_base
+from theseus.labs.lie.functional import so3 as _so3_base
+
+
+TensorType = Union[torch.Tensor, "LieTensor"]
+
+
+class LieType(Enum):
+    SE3 = 0
+    SO3 = 1
+
+
+SE3 = LieType.SE3
+SO3 = LieType.SO3
+
+_JFnReturnType = Tuple[List[torch.Tensor], TensorType]
+
+
+class LieTensor:
+    def __init__(self, data_tensor: torch.Tensor, ltype: LieType):
+        self._t = data_tensor
+        self._fn_lib = {
+            SE3: _se3_base,
+            SO3: _so3_base,
+        }[ltype]
+        self._fn_lib.check_group_tensor(data_tensor)
+        self.ltype = ltype
+
+    def __repr__(self) -> str:
+        return f"LieTensor({self._t}, ltype=lie.{self.ltype})"
+
+    def _check_ltype(self, other: "LieTensor", op_name: str):
+        if other.ltype != self.ltype:
+            raise ValueError("f{op_name} requires both tensors to have same ltype.")
+
+    # Returns a new LieTensor with the given data and the same ltype as self
+    def new(self, t: TensorType) -> "LieTensor":
```

Comment:
Maybe also a `as_lietensor`?

Reply:
I was thinking about something like `torch.Tensor.new_tensor(x)`, which uses the data in `x` but makes a tensor with the same `dtype` and `device`.

Reply:
`self` is used to pass the `ltype` along (line 49).

Reply:
```python
x = lie.LieTensor(xdata, lie.LieType.SE3)
t = torch.Tensor(tdata)

y = x.new(x) # Is this usage a bit strange?
z = x.new(t) # this seems okay

# on the other hand both below are fine
y = lie.new(x)
z = lie.new(t) # for this maybe ltype can be passed to new as an optional argument?
```

Reply:
Actually `lie.new(t, lie.LieType.SE3)` is basically same as `lie.LieTensor()`. I guess we can keep as is (or make new only support LieTensor not tensor)?

Reply:
`z = x.new(y)` is meant to receive new data from y, but copy dtype, ltype, device (and other tensor properties) from x. In `torch` one could do `y = x.new_tensor(x)`, but this would be a pointless operation (`torch` throws a warning to use `clone()` instead).

Reply:
Also, not a fan of `lie.LieType`. Torch simply refers to `torch.double`, `torch.float`, etc. Why not follow this convention?

Reply:
Okay cool, I like the idea to follow torch here with the clone warning (maybe we can add a clone op? sorry for dragging this out too much).

Ah ignore my bad use of LieType, haha. Yes we should have it setup like you suggest with lie.SE3, etc.

Reply:
Actually, after playing around with this, I think this should have just being 'clone()' all along :) 

Following torch, `x.new()` would be more appropriate if one were to pass an arbitrary shape. For example in `torch` one can do `x.new_tensor([1, 2, 3]))`, which makes a new tensor of shape (3,) with the contents in the list.

Reply:
In the end decided to keep `new()` as well, since we are using to create a LieTensor from a torch tensor. I also added clone and a warning. Let me know if it looks OK.

Reply:
Looks good!

Reply:
Not sure if this affects the compatibility of functorch and `autograd.jacobian` with `vectorize=True`, in both cases the `tensor` is not a regular pytorch tensor.